### PR TITLE
Attrs as instance vars

### PR DESCRIPTION
**Pull request by santib**

### CODE REVIEW

File: lib/ar2dto/dto.rb [hunk 1]
```diff
@@ -6,17 +6,22 @@ def self.[](original_model)
...
+        attr_reader(*original_model.attribute_names)
       end
     end
 
-    def initialize(attributes = {})
-      attributes.each { |key, value| define_singleton_method(key) { value } }
-      super()
-    end
-
-    def as_json(options = nil)
```
Comment (sebaherrera07) [inline lib/ar2dto/dto.rb:17]: Is this method no longer necessary?

File: spec/ar2dto/dto_spec.rb [hunk 2]
```diff
@@ -49,7 +49,7 @@ def self.name
     end
 
     context "with options" do
-      let(:options) { { except: :first_name } }
+      let(:options) { { except: "first_name" } }
```
Comment (sebaherrera07) [inline spec/ar2dto/dto_spec.rb:52]: Just curious, why these changes happened?

Shouldn't the options be passed as symbols?

File: spec/ar2dto/options_spec.rb [hunk 3]
```diff
@@ -22,12 +32,12 @@ def self.name
           has_dto
         end
 
-        expect(klass.new.to_dto).to_not respond_to(:updated_at)
+        expect(klass.new(attributes).to_dto.updated_at).to be_nil
```
Comment (sebaherrera07) [inline spec/ar2dto/options_spec.rb:35]: Is this behavior desired?

Wouldn't we prefer to make it not respond to that message as it was before?

File: lib/ar2dto/dto.rb (hunk 1 above)
Comment (santib) [reply lib/ar2dto/dto.rb:17]: right, `as_json` is defined by Rails on objects by building a hash with the instance vars that's exactly what we want

File: spec/ar2dto/dto_spec.rb (hunk 2 above)
Comment (santib) [reply spec/ar2dto/dto_spec.rb:52]: good question, we probably need to support symbols as well, I'll check

File: spec/ar2dto/options_spec.rb (hunk 3 above)
Comment (santib) [reply spec/ar2dto/options_spec.rb:35]: yes maybe that's better. I wanted to define the attr_readers at the class level with all the attributes but might be a bit strange having these methods return nil. I'll try the other approach

Comment (sebaherrera07) [reply spec/ar2dto/options_spec.rb:35]: Agree, because if the DTO responds to these messages and returns `nil`, some people might in some cases wrongfully use the returned value as if it was the actual value.

File: spec/ar2dto/dto_spec.rb (hunk 2 above)
Comment (santib) [reply spec/ar2dto/dto_spec.rb:52]: just checked, this is the `as_json` Rails method, so I think it's fine

File: spec/ar2dto/dto_spec.rb [hunk 4]
```diff
@@ -29,7 +29,7 @@ def self.name
 
     it "implements equality" do
       expect(subject).to eq(klass.new(user.attributes))
-      expect(subject).to_not eq(klass.new(user.attributes.merge(first_name: "Other")))
+      expect(subject).to_not eq(klass.new(user.attributes.merge("first_name" => "Other")))
```
Comment (martinjaimem) [inline spec/ar2dto/dto_spec.rb:32]: Can we actually add a test to check that it is an instance variable?

```ruby
some_object.instance_variable_get("@first_name")
```

**Review by martinjaimem — APPROVED** [review]
